### PR TITLE
Adds `drake_gym` to Drake for OpenAI Gym compatibility

### PR DESCRIPTION
**Pull request by ggould-tri**
Towards #15508.

This commit brings Drake Gym into Drake.  This is a tool for turning a Drake Diagram into an OpenAI Gym Env.  Two examples are provided to clarify this rather complex operation.

Note:  This commit includes `gym` and a patched version of `stable_baselines3` to allow tests to run.  However this is not a full version of `stable_baselines3`; users who want to use that will need to install it themselves.

- - -
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18178)

### CODE REVIEW

**Review by ggould-tri — COMMENTED** [review]
+(status: do not merge) +(status: do not review) 

@jwnimmer-tri This is a draft pull request that I'm using for cleaning/checklisting; however, it shows the patch to sb3 that I used to nerf its dependencies enough to use check_env.  I don't think at the moment there is any plausible hope of getting any of the optimizers working without torch, so the training scripts are conditionalized into inactivity for now.

I will eagerly accept any suggestions, corrections, or outright edits of any of the workspace matter, in which I have no confidence  (or indeed of anything else, but you're not likely to find the rest very edifying).

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18178#-:-NFFICpn9uCiC2Y9tsnE:b9vmts8)* status: needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge", missing label for release notes

**Review by jwnimmer-tri — COMMENTED** [review]
I did a high-level read of the workspace stuff, with no real problems with what's here so far.  I posted a few nits, but I'll open one new top-level discussion thread about my only open question.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18178#-:-NFFSsXd7-hRPXcKCTEy:bfqpe1b)* status: 5 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge", missing label for release notes (waiting on @ggould-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFTqre2jiAwAROfE3d:-NFFTqre2jiAwAROfE3e:befi18f) (no related file):*
Does the lack of any non-pytorch optimizers mean that most (or all) of the examples here are not sufficiently exercised in CI?

If yes, then we need to come up with some answer to get CI coverage of those.  In the extreme, we could have drake-external-examples CI run them, using a GitHub Dockerfile that has torch already in it (hopefully that's a thing which already exists in the universe somewhere).

___
___
*[`drake_gym/drake_gym.py` line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFR7i--3RxG81kj1vG:-NFFR7i--3RxG81kj1vH:b60lry2) ([raw file](https://github.com/robotlocomotion/drake/blob/8f2b68911798355eb966cfb8e4fed25024b9c3c7/drake_gym/drake_gym.py#L1)):*
> ```Python
> from typing import Callable, Optional, Union
> ```

nit If we get traction this PR (versus a separate repo), then we might as well put this into is final home right away (`drake/bindings/pydrake/gym/__init__.py`).

When it's here instead of within pydrake, it won't be usable outside of bazel.
___
*[`setup/ubuntu/source_distribution/packages-focal-test-only.txt` line 5 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFT7rX1k8T9lM3Ho8Z:-NFFT7rX1k8T9lM3Ho8_:b-4sztc3) ([raw file](https://github.com/robotlocomotion/drake/blob/8f2b68911798355eb966cfb8e4fed25024b9c3c7/setup/ubuntu/source_distribution/packages-focal-test-only.txt#L5)):*
> ```help
> libstdc++6-10-dbg
> llvm-12
> python3-cloudpickle
> ```

nit This needs a likewise mention in `drake/setup/mac/source_distribution/requirements-test-only.txt`.
___
*[`tools/workspace/gym_py/package.BUILD.bazel` line 7 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFRqZpCe76CDs3YGVT:-NFFRqZpCe76CDs3YGVU:bnimh9u) ([raw file](https://github.com/robotlocomotion/drake/blob/8f2b68911798355eb966cfb8e4fed25024b9c3c7/tools/workspace/gym_py/package.BUILD.bazel#L7)):*
> ```
> licenses(["notice"])  # MIT
> 
> # Internal vendoring of a few files of stable_baselines3; see README.md
> ```

nit This line seems stale.
___
*[`tools/workspace/stable_baselines3_internal/no_optimizers.patch` line 103 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFSZje-1vspGSiNYZ9:-NFFSZje-1vspGSiNYZA:b-cwmd6t) ([raw file](https://github.com/robotlocomotion/drake/blob/8f2b68911798355eb966cfb8e4fed25024b9c3c7/tools/workspace/stable_baselines3_internal/no_optimizers.patch#L103)):*
> ```Diff
>  import multiprocessing as mp
>  from collections import OrderedDict
> +from multiprocessing import connection
> ```

nit Is this just a bug in upstream, unrelated to our remixing (seems like yes)?  If so, then it should be a distinct `*.patch` file to clarify that purpose, and we should attempt to submit it upstream for eventual inclusion.

**Review by ggould-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18178)* status: 5 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge", missing label for release notes (waiting on @jwnimmer-tri)

___
*[`tools/workspace/stable_baselines3_internal/no_optimizers.patch` line 103 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFSZje-1vspGSiNYZ9:-NFFUkFf0urGoQCEt29T:br64fee) ([raw file](https://github.com/robotlocomotion/drake/blob/8f2b68911798355eb966cfb8e4fed25024b9c3c7/tools/workspace/stable_baselines3_internal/no_optimizers.patch#L103)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit Is this just a bug in upstream, unrelated to our remixing (seems like yes)?  If so, then it should be a distinct `*.patch` file to clarify that purpose, and we should attempt to submit it upstream for eventual inclusion.
</blockquote></details>

AFAICT this doesn't fail upstream, probably due to a transitive dependency leaking in from a __init__.py or something.  I am not at all sure what made it necessary here.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18178)* status: 4 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge", missing label for release notes (waiting on @ggould-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFTqre2jiAwAROfE3d:-NFFVjRqAyJ2CR8zFPIR:b4b6bkz) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Does the lack of any non-pytorch optimizers mean that most (or all) of the examples here are not sufficiently exercised in CI?

If yes, then we need to come up with some answer to get CI coverage of those.  In the extreme, we could have drake-external-examples CI run them, using a GitHub Dockerfile that has torch already in it (hopefully that's a thing which already exists in the universe somewhere).
</blockquote></details>

For the tests, the other option we have at hand is instead of patching sb3 to avoid the imports, we can insert stub implementations of the modules onto the path so that (some of) the imports succeed.

We already have an example of that for [torch](https://github.com/RobotLocomotion/drake/blob/master/bindings/pydrake/test/mock_torch/torch.py) and [scipy](https://github.com/RobotLocomotion/drake/blob/master/bindings/pydrake/common/test_utilities/scipy_stub/scipy/sparse/__init__.py).

I wonder how much of `torch` we actually need to run one of the simpler optimizer.  If it's just torch.min = np.min or some boilerplate, maybe its viable.

___

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18178)* status: 4 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge", missing label for release notes (waiting on @ggould-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFTqre2jiAwAROfE3d:-NFFcB2Z5wGvFNvlLNyk:b-hkacny) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

For the tests, the other option we have at hand is instead of patching sb3 to avoid the imports, we can insert stub implementations of the modules onto the path so that (some of) the imports succeed.

We already have an example of that for [torch](https://github.com/RobotLocomotion/drake/blob/master/bindings/pydrake/test/mock_torch/torch.py) and [scipy](https://github.com/RobotLocomotion/drake/blob/master/bindings/pydrake/common/test_utilities/scipy_stub/scipy/sparse/__init__.py).

I wonder how much of `torch` we actually need to run one of the simpler optimizer.  If it's just torch.min = np.min or some boilerplate, maybe its viable.
</blockquote></details>

Here's a sufficient stub patch for the unit test (check env) to pass: https://github.com/jwnimmer-tri/drake/commit/34e9acd791390a931c85edc4a00469435636cfe5

Take a look and see what you think, but my guess is that maintaining enough of `import torch` to typecheck the sb3 imports will be easier over time than patching sb3 directly.

___

**Comment by ggould-tri** [comment]
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Here's a sufficient stub patch for the unit test (check env) to pass: https://github.com/jwnimmer-tri/drake/commit/34e9acd791390a931c85edc4a00469435636cfe5

Take a look and see what you think, but my guess is that maintaining enough of `import torch` to typecheck the sb3 imports will be easier over time than patching sb3 directly.
</blockquote></details>

Nice!  That's a much cleaner path to getting `drake_gym_test` to pass.
For getting coverage through the `--test` invocations of the training, we'd need the types (and possibly functional bits) from `torch` which is probably harder, but I'll look into it.

**Review by ggould-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18178)* status: 2 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge", missing label for release notes (waiting on @ggould-tri and @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFTqre2jiAwAROfE3d:-NGYBSt55yAgqi4dkJoa:b-pfm64j) (no related file):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Nice!  That's a much cleaner path to getting `drake_gym_test` to pass.
For getting coverage through the `--test` invocations of the training, we'd need the types (and possibly functional bits) from `torch` which is probably harder, but I'll look into it.
</blockquote></details>

Incorporated your branch; still looking into the rest.

___
___
*[`drake_gym/drake_gym.py` line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFR7i--3RxG81kj1vG:-NGYBXkM-JGKAHkK6mo8:bq0efpf) ([raw file](https://github.com/robotlocomotion/drake/blob/8f2b68911798355eb966cfb8e4fed25024b9c3c7/drake_gym/drake_gym.py#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit If we get traction this PR (versus a separate repo), then we might as well put this into is final home right away (`drake/bindings/pydrake/gym/__init__.py`).

When it's here instead of within pydrake, it won't be usable outside of bazel.
</blockquote></details>

Working on this.

**Review by ggould-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18178)* status: 2 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge", commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @ggould-tri and @jwnimmer-tri)

___
*[`drake_gym/drake_gym.py` line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/18178#-NFFR7i--3RxG81kj1vG:-NGreR2vAFlPy9J5w-Vr:bwhglds) ([raw file](https://github.com/robotlocomotion/drake/blob/8f2b68911798355eb966cfb8e4fed25024b9c3c7/drake_gym/drake_gym.py#L1)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Working on this.
</blockquote></details>

Moved.

**Comment by jwnimmer-tri** [comment]
Shall we close this in lieu of https://github.com/RobotLocomotion/drake/pull/19831?
